### PR TITLE
feat(EnhancedClient): Detect and Error on Ignored DynamoDB Encryption Tags

### PR DESCRIPTION
**Pull request by texastony**
*Issue #, if available:* Detect and Error on Ignored DynamoDB Encryption Tags

*Description of changes:*
- Tests using Classes that use Annotations from `awssdk.enhanced.dynamodb.mapper` to define nested object mapping
- So far, the Annotations tested are:
  - `@DynamoDbConvertedBy`
  - `@DynamoDbFlatten`
  - straight up nested object with no annotation
  - Conflicting Annotations
- Refactor `DynamoDbEnhancedClientEncryption` to identify some IGNORED DynamoDB Encryption Tags and throw an Exception when they are encountered.
- https://github.com/aws/aws-database-encryption-sdk-dynamodb-java/pull/252 Documents two shallow configurations with ignored annotations that are NOT detected by this implementation. Only of these is of any real concern: [`ConflictingFlattenedBean.java`](https://github.com/aws/aws-database-encryption-sdk-dynamodb-java/blob/2d26691e0ff8cd10dcaf95aa2f6d4ba020e8495f/DynamoDbEncryption/runtimes/java/src/test/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/ConflictingFlattenedBean.java)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java [hunk 1]
```diff
@@ -57,6 +60,8 @@ private static DynamoDbTableEncryptionConfig getTableConfig(DynamoDbEnhancedTabl
                 // non-key attributes are ENCRYPT_AND_SIGN unless otherwise annotated
                 actions.put(attributeName, CryptoAction.ENCRYPT_AND_SIGN);
             }
+
```
Comment (seebees) [inline DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:63]: Lets add a comment here explaining why this is here

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java [hunk 2]
```diff
@@ -90,4 +95,41 @@ private static DynamoDbTableEncryptionConfig getTableConfig(DynamoDbEnhancedTabl
                 .legacyOverride(configWithSchema.legacyOverride())
                 .build();
     }
+
+    /**
+     * Detects DynamoDB Encryption Tags in Nested Enhanced Types.<p>
```
Comment (seebees) [inline DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:100]: Add comment about why flatten "works".

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java [hunk 3]
```diff
@@ -90,4 +95,41 @@ private static DynamoDbTableEncryptionConfig getTableConfig(DynamoDbEnhancedTabl
...
+    ) {
+        AttributeConverter attributeConverter = configWithSchema.schemaOnEncrypt().converterForAttribute(attributeName);
+        if (
+            Objects.nonNull(attributeConverter) &&
+                Objects.nonNull(attributeConverter.type()) &&
+                attributeConverter.type().tableSchema().isPresent()
+        ) {
+            Object maybeTableSchema = attributeConverter.type().tableSchema().get();
+            if (maybeTableSchema instanceof TableSchema) {
+                TableSchema subTableSchema = (TableSchema) maybeTableSchema;
```
Comment (seebees) [inline DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:120]: I think that having a `for (String attributeName : attributeNames) {` for this subTableSchema would turn this into a deep scanner right?

Comment (seebees) [reply DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:120]: This relies on DDD not accepting a recursive data structure

**Review by seebees — DISMISSED** [review]

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java (hunk 1 above)
Comment (texastony) [reply DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:88]: Resolved in 90ac3d70

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java (hunk 2 above)
Comment (texastony) [reply DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:126]: Resolved in 90ac3d70

File: DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java (hunk 3 above)
Comment (texastony) [reply DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:153]: We will make a new PR and GHI to track this.
I agree that it SHOULD just work,
but we need to author test cases and implement it...

Comment (texastony) [reply DynamoDbEncryption/runtimes/java/src/main/java/software/amazon/cryptography/dbencryptionsdk/dynamodb/enhancedclient/DynamoDbEnhancedClientEncryption.java:153]: GHI tracking this is https://github.com/aws/aws-database-encryption-sdk-dynamodb-java/issues/258

**Review by ajewellamz — APPROVED** [review]

**Pull request merged**